src/Disciplina/WorldState/Internal.hs

```diff
@@ -0,0 +1,495 @@
+      where
+        (==?) :: (Show s, Eq s, Binary s) => Proof s -> Proof s -> Bool
+        (==?) = (==) `on` getHash
+
+        -- TODO: move into Data.Tree.AVL.Proof
+        getHash proof = proof
+            ^.AVL._Proof    -- get the tree
+             .AVL.rootHash  -- reduce to hash
+
+type Proof = AVL.Proof Hash Entity
+
+type DAG         = Hash
+type Publication = Hash
+type Storage     = Hash
+type Code        = Hash
+
+-- | The changes the transaction consists from.
+data Change
+    = TransferTokens Entity      Amount
+    | Publicate      Publication
+    | CreateAccount  Entity      Code
+    deriving (Show)
+
+-- TODO: implement signing (and make 'Entity' to be actual public key).
+data Transaction = Transaction
+    { _tAuthor     :: Entity
+    , _tChanges    :: [Change]
+    , _tNonce      :: Int
+    }
+
+instance Show Transaction where
+    show (Transaction who what nonce) =
+        concat
+            [ show who
+            , ": "
+            , show what
+            , " ("
+            , show nonce
+            , ")"
+            ]
+
+-- | Wrapper that adds proof to either block or transaction.
+data WithProof a = WithProof
+    { _wpBody    :: a
+    , _wpProof   :: WorldStateProof
+    , _wpEndHash :: Hash
+    }
+
+-- | World state on server.
+data Server = Server
+    { _sWorld :: WorldState
+    }
+
+-- | World state on client.
+data Client = Client
+    { _cProof :: WorldStateProof
+    }
+
+-- | Set of node identities to generate proof from. Proof prefab.
+data RevisionSets = RevisionSets
+    { _accountRevSet       :: AVL.RevSet
+    , _publicationRevSet   :: AVL.RevSet
+    , _specalizationRevSet :: AVL.RevSet
+    , _storageRevSet       :: AVL.RevSet
+    , _codeRevSet          :: AVL.RevSet
+    }
+    deriving (Default, Generic)
+
+data Environment = Environment
+    { _eAuthor :: Entity
+    }
+
+-- | Variants of failure cases.
+data TransactionError
+    = AccountDoesNotExist       { doesNotExist   :: Sided Entity }
+    | AccountExistButShouldNot  { shouldNotExist :: Sided Entity }
+
+    | NotEnoughTokens           { holder         :: Entity
+                                , wantsToSpend   :: Amount
+                                , hasOnly        :: Amount
+                                }
+    | NoncesMismatch            { entity         :: Entity
+                                , hasNonce       :: Int
+                                , requiredNonce  :: Int
+                                }
+    | InitialHashesMismatch
+    | FinalHashesMismatch
+    deriving (Show, Typeable)
+
+deriving instance Exception TransactionError
+
+-- | Enrich 'Entity' with its 'Side' in the deal to improve error messages.
+data Sided a = Sided
+    { who  :: a
+    , side :: Side
+    }
+    deriving (Show, Typeable)
+
+data Side = Sender | Receiver
+    deriving (Show, Typeable)
+
+-- | Monad for operations.
+type WorldM side = RWST Environment RevisionSets side IO
```

Comment:
Side is either a server state (full trees) of a client one (maximally reduced trees).

The revision sets, logged during an operation, are used to construct a proof of that operation.

I prefer using `TChan`s for processes to communicate. I see no reason to have `TVar` here. Do we have concurrent writes? No, the source of updates is singular. In the end of the day, the threads (persistence, whatever) depending on that information will only need to read it when it changes once per block arrived.

Reply:
>Side is either a server state (full trees) of a client one (maximally reduced trees).
Got the idea. But I don't think that it's really time to think about light clients right now.

>The revision sets, logged during an operation, are used to construct a proof of that operation.
Wait, but aren't AVL+ trees operations construct proofs simultaneously with operation itself?

>Do we have concurrent writes?
Oh, for sure we'll have. Processing a newly created transaction simultaneously with processing newly created block is quite common. Also concurrent read/write operations are going to be ubiquitous.

>In the end of the day, the threads (persistence, whatever) depending on that information will only need to read it when it changes once per block arrived.
You forgot about the mempool updates, for example.

Of course, we can think about only using `TChan`s. It would be even better to think about FRP. The point is: we _already_ know how to make things in a simple way, with `TVar`s and stuff, which makes it safe to start writing code this way. But if we try to do better without proper thinking before, we can collect all the perils and lose lots of time.